src/coprocessor/Client.cc

```diff
@@ -35,24 +44,378 @@ std::vector<copTask> buildCopTasks(
         // all ranges belong to same region.
         if (i == ranges.size())
         {
-            tasks.push_back(copTask{loc.region, ranges, cop_req, store_type, meta_data, before_send});
+            tasks.push_back(CopTask{loc.region, ranges, cop_req, store_type, /*partition_index=*/0, meta_data, before_send});
             break;
         }
-        std::vector<KeyRange> task_ranges(ranges.begin(), ranges.begin() + i);
+
+        KeyRanges task_ranges(ranges.begin(), ranges.begin() + i);
+        // Split the last range if it is overlapped with the region
         auto & bound = ranges[i];
         if (loc.contains(bound.start_key))
         {
             task_ranges.push_back(KeyRange{bound.start_key, loc.end_key});
-            bound.start_key = loc.end_key;
+            bound.start_key = loc.end_key; // update the last range start key after splitted
         }
-        tasks.push_back(copTask{loc.region, task_ranges, cop_req, store_type, meta_data, before_send});
+        tasks.push_back(CopTask{loc.region, task_ranges, cop_req, store_type, /*partition_index=*/0, meta_data, before_send});
         ranges.erase(ranges.begin(), ranges.begin() + i);
     }
     log->debug("has " + std::to_string(tasks.size()) + " tasks.");
     return tasks;
 }
 
-std::vector<copTask> ResponseIter::handleTaskImpl(kv::Backoffer & bo, const copTask & task)
+
+namespace details
+{
+std::vector<LocationKeyRanges> splitKeyRangesByLocations(
```

Comment:
If we move this function into RegionCache, then `RegionCache.h` need to include `Client.h` for the definition in `struct KeyRange`. It will easily cause header file loops including.
Maybe we should use a separated file to define `struct KeyRange` and `struct LocationKeyRanges`